parser/parser/cc/grammars/typedruby.ypp

```diff
@@ -927,28 +928,71 @@ int yylex(parser::semantic_type *lval, ruby_parser::location *lloc, ruby_parser:
                       driver.lex.context.allowNumparams = false;
                     }
 
-        defn_head: k_def def_name
+       defn_head: k_def def_name
                     {
                       auto defNameToken = $2->token;
+                      ForeignPtr node = nullptr;
+                      if (driver.rewind_if_different_line($1, defNameToken)) {
+                        node = driver.build.defnHeadError(self, $1);
+                      } else {
+                        node = driver.build.defnHead(self, $1, defNameToken);
```

Comment:
Good question, and no, because the guard at the top of `rewind_if_different_line` exits early on the first parse (when `indentationAware` is false).

It does mean that our heuristic could recover in a weird way in a file that already had a syntax error though. For example:

```ruby
class A
  def
    foo
  end
  def bar
end
```

The changes in this PR recover from that in a weird way:

```ruby
class A
  def
    foo
  end
  def bar
end
```

We're ~okay with this (it could obviously be better) because this heuristic only applies when the first parse already produced no result, so it doesn't exactly matter what specifically we produce in the name of recovery, with the assumption being that something is better than nothing.